service/history/workflow/mutable_state_impl.go

```diff
@@ -5263,6 +5301,64 @@ func (ms *MutableStateImpl) closeTransactionUpdateTransitionHistory(
 	return nil
 }
 
+func (ms *MutableStateImpl) closeTransactionTrackLastUpdatedVersionedTransition(
+	transactionPolicy TransactionPolicy,
+) error {
+	if transactionPolicy != TransactionPolicyActive {
+		// TODO: replication/standby logic will need a different way for updating LastUpdatedVersionedTransition
+		// when reapplying history, especially when history replication tasks got batched.
+		return nil
+	}
+
+	if len(ms.executionInfo.TransitionHistory) == 0 {
+		// transition history is not enabled
+		return nil
+	}
+
+	transitionHistory := ms.executionInfo.TransitionHistory
+	currentVersionedTransition := transitionHistory[len(transitionHistory)-1]
+	for _, activityInfo := range ms.updateActivityInfos {
+		activityInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+	for _, timerInfo := range ms.updateTimerInfos {
+		timerInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+	for _, childInfo := range ms.updateChildExecutionInfos {
+		childInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+	for _, cancelInfo := range ms.updateRequestCancelInfos {
+		cancelInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+	for _, signalInfo := range ms.updateSignalInfos {
+		signalInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+
+	// signalRequestedIDs is a set in DB, we don't have a place to store the lastUpdateVersionedTransition for each
+	// signal requestedID.
+	// Deletion of signalRequestID is not replicated today, so we can even drop the check on deleteSignalRequestedIDs
+	if len(ms.updateSignalRequestedIDs) != 0 || len(ms.deleteSignalRequestedIDs) != 0 {
+		ms.executionInfo.SignalRequestIdsLastUpdatedVersionedTransition = currentVersionedTransition
+	}
+
+	for updateID := range ms.updatedUpdateInfosVersionedTransition {
+		ms.executionInfo.UpdateInfos[updateID].LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+
+	if ms.updateWorkflowTaskVersionedTransition && ms.HasPendingWorkflowTask() {
+		ms.executionInfo.WorkflowTaskLastUpdatedVersionedTransition = currentVersionedTransition
+	}
+
+	if ms.updateVisibilityVersionedTransition {
+		ms.executionInfo.VisibilityLastUpdatedVersionedTransition = currentVersionedTransition
```

Comment:
An alternative implementation could be check if there's any visibility task generated. However that couples the logic with task generation, which could happen due to say task refresh but that's not a actual transition, so I don't like that approach.